File: adapters/telemetry/wire.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Wire glue — a reading <-> a self-framing Fibonacci bitstream, through the decision-preserving quantizer.

Field order is canonical (sorted field names), so encode and decode agree with **zero header**. Symbols
are 0-based cell indices; the Fibonacci codec is defined for positive integers, so each is sent as
``symbol + 1``. The composition is the whole point: quantize (decision-preserving) -> Fibonacci-code
(self-framing) -> decode -> reconstruct routes identically to the original reading.
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from prismpath import predicates

import quantizer as q
import zeckendorf as z


def _order(parts: Dict[str, "q.FieldPartition"]) -> List[str]:
    return sorted(parts.keys())


def encode_reading(parts: Dict[str, "q.FieldPartition"], reading: Dict[str, Any]) -> str:
    """Reading -> bitstream. Requires every decision-relevant field present (missing-field handling is a
    later refinement)."""
    order = _order(parts)
    missing = [f for f in order if f not in reading]
    if missing:
        raise KeyError(f"reading missing decision fields: {missing}")
    syms = q.quantize(parts, reading)
    return z.encode_stream([syms[f] + 1 for f in order])


def decode_reading(parts: Dict[str, "q.FieldPartition"], bits: str) -> Dict[str, Any]:
    """Bitstream -> a representative reading that routes identically to the original."""
    order = _order(parts)
    wire = z.decode_stream(bits)
    if len(wire) != len(order):
        raise ValueError(f"symbol count {len(wire)} != decision fields {len(order)}")
    syms = {f: wire[i] - 1 for i, f in enumerate(order)}
    return q.reconstruct(parts, syms)


# --------------------------------------------------------- reference routing (the deterministic tier)
def route_node(graph, node: str, reading: Dict[str, Any]) -> Optional[str]:
    """First-match deterministic routing from `node` over a reading — the engine's field-routing tier."""
    for target, cond in graph.nodes[node].edges:
        if predicates.is_deterministic(cond) and predicates.eval_condition(cond, reading):
            return target
    return None


def decision_nodes(graph) -> List[str]:
    """Nodes with at least one deterministic (field-routing) edge."""
    return [name for name, node in graph.nodes.items()
            if any(predicates.is_deterministic(c) for _t, c in node.edges)]

File: prismpath-telemetry-rs/src/wire.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! Wire glue — a reading <-> a self-framing Fibonacci bitstream, through the decision-preserving quantizer.

use crate::quantizer::{self, FieldPartition};
use crate::zeckendorf;
use prismpath_rs::{Graph, V};
use std::collections::HashMap;

pub trait WireCodec {
    fn encode_symbols(&self, symbols: &[usize]) -> Result<String, String>;
    fn decode_symbols(&self, bits: &str) -> Vec<usize>;
}

#[derive(Debug, Clone, Copy, Default)]
pub struct FibonacciWireCodec;

impl WireCodec for FibonacciWireCodec {
    fn encode_symbols(&self, symbols: &[usize]) -> Result<String, String> {
        zeckendorf::encode_stream(symbols)
    }

    fn decode_symbols(&self, bits: &str) -> Vec<usize> {
        zeckendorf::decode_stream(bits)
    }
}

pub fn order(parts: &HashMap<String, FieldPartition>) -> Vec<String> {
    let mut keys: Vec<String> = parts.keys().cloned().collect();
    keys.sort();
    keys
}

pub fn encode_reading(
    parts: &HashMap<String, FieldPartition>,
    reading: &HashMap<String, V>,
) -> Result<String, String> {
    encode_reading_with_codec(&FibonacciWireCodec, parts, reading)
}

pub fn encode_reading_with_codec<C: WireCodec>(
    codec: &C,
    parts: &HashMap<String, FieldPartition>,
    reading: &HashMap<String, V>,
) -> Result<String, String> {
    let ord = order(parts);
    let missing: Vec<&str> = ord
        .iter()
        .filter(|f| !reading.contains_key(*f))
        .map(|s| s.as_str())
        .collect();
    if !missing.is_empty() {
        return Err(format!("reading missing decision fields: {:?}", missing));
    }
    let syms = quantizer::quantize(parts, reading)?;
    let wire_vals: Vec<usize> = ord.iter().map(|f| syms[f] + 1).collect();
    codec.encode_symbols(&wire_vals)
}

pub fn decode_reading(
    parts: &HashMap<String, FieldPartition>,
    bits: &str,
) -> Result<HashMap<String, V>, String> {
    decode_reading_with_codec(&FibonacciWireCodec, parts, bits)
}

pub fn decode_reading_with_codec<C: WireCodec>(
    codec: &C,
    parts: &HashMap<String, FieldPartition>,
    bits: &str,
) -> Result<HashMap<String, V>, String> {
    let ord = order(parts);
    let wire = codec.decode_symbols(bits);
    if wire.len() != ord.len() {
        return Err(format!(
            "symbol count {} != decision fields {}",
            wire.len(),
            ord.len()
        ));
    }
    let mut syms: HashMap<String, usize> = HashMap::new();
    for (i, f) in ord.into_iter().enumerate() {
        let s = wire[i]
            .checked_sub(1)
            .ok_or_else(|| format!("invalid wire symbol 0 for field {f}"))?;
        syms.insert(f, s);
    }
    Ok(quantizer::reconstruct(parts, &syms))
}

pub fn route_node(graph: &Graph, node: &str, reading: &HashMap<String, V>) -> Option<String> {
    if let Some(n) = graph.nodes.get(node) {
        for (target, cond) in &n.edges {
            if prismpath_rs::is_deterministic(cond) {
                if let Ok(true) = prismpath_rs::eval_condition(cond, reading) {
                    return Some(target.clone());
                }
            }
        }
    }
    None
}

pub fn decision_nodes(graph: &Graph) -> Vec<String> {
    decision_nodes_from_text(graph, None)
}

pub fn decision_nodes_from_text(graph: &Graph, text: Option<&str>) -> Vec<String> {
    let mut names = Vec::new();
    if let Some(txt) = text {
        for line in txt.lines() {
            let l = line.trim();
            if let Some(rest) = l.strip_prefix("## ") {
                let name = rest.trim().to_lowercase().replace(' ', "_");
                if let Some(node) = graph.nodes.get(&name) {
                    if node.edges.iter().any(|(_t, c)| prismpath_rs::is_deterministic(c))
                        && !names.contains(&name)
                    {
                        names.push(name.clone());
                    }
                }
            }
        }
    }
    if names.is_empty() {
        let mut ordered_keys = Vec::new();
        if graph.nodes.contains_key(&graph.start) {
            ordered_keys.push(graph.start.clone());
        }
        let mut rest: Vec<String> = graph.nodes.keys().cloned().collect();
        rest.sort();
        for k in rest {
            if !ordered_keys.contains(&k) {
                ordered_keys.push(k);
            }
        }
        for name in ordered_keys {
            if let Some(node) = graph.nodes.get(&name) {
                if node.edges.iter().any(|(_t, c)| prismpath_rs::is_deterministic(c)) {
                    names.push(name);
                }
            }
        }
    }
    names
}
